From a5934496525e58ff6cf840464bc9938167a56c12 Mon Sep 17 00:00:00 2001
From: Xiaolei Wang <xiaolei.wang@windriver.com>
Date: Mon, 25 Nov 2019 16:45:17 +0800
Subject: [PATCH] nxp-imx6: fix a Memory leak bug in mxsfb_init_fbinfo_dt

unreferenced object 0x88c09e40 (size 64):
  comm "swapper/0", pid 1, jiffies 4294937853 (age 205.150s)
  hex dump (first 32 bytes):
    00 ff bb 88 00 00 00 00 00 00 00 00 00 00 00 00  ................
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
  backtrace:
    [<17c4f4a3>] kmemleak_alloc+0x40/0x74
    [<08f1aa43>] __kmalloc+0x240/0x304
    [<376d1947>] of_get_display_timings+0x114/0x204
    [<1a5d2b4a>] mxsfb_init_fbinfo_dt+0x1a4/0x2cc
    [<ccef2d42>] mxsfb_probe+0x2e8/0xd28
    [<d8df3f12>] platform_drv_probe+0x58/0xa4
    [<5ce7fb0e>] driver_probe_device+0x240/0x308
    [<e300ddf1>] __driver_attach+0x9c/0xd8
    [<8e15a7a4>] bus_for_each_dev+0x78/0xb8
    [<42f6adad>] driver_attach+0x28/0x30
    [<913ca58e>] bus_add_driver+0x184/0x1ec
    [<7d17daa1>] driver_register+0xbc/0x100
    [<0124e09c>] __platform_driver_register+0x40/0x54
    [<3b8d3955>] mxsfb_driver_init+0x20/0x28
    [<66ef1f6d>] do_one_initcall+0xf8/0x264
    [<6a2fa56d>] kernel_init_freeable+0x354/0x440

In the mxsfb_init_fbinfo_dt function, of_get_display_timings applies
memory for timings and timings members,so you should call
display_timings_release and it`s member timings memory instead of
calling kfree,which will cause memory leaks.

Signed-off-by: Xiao Wang <Xiaolei.Wang@windriver.com>
---
 drivers/video/fbdev/mxsfb.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/video/fbdev/mxsfb.c b/drivers/video/fbdev/mxsfb.c
index b28c3b9c6b24..ab402d3cc60e 100644
--- a/drivers/video/fbdev/mxsfb.c
+++ b/drivers/video/fbdev/mxsfb.c
@@ -1410,7 +1410,7 @@ static int mxsfb_init_fbinfo_dt(struct fb_info *fb_info)
 	of_node_put(timings_np);
 put_display_node:
 	if (timings)
-		kfree(timings);
+		display_timings_release(timings);
 	of_node_put(display_np);
 	return ret;
 }
-- 
2.17.1

